File: rings.go
package main

import (
	"fmt"
)

// Функция для ввода строки с кольцами
func input() string {
	var rings string
	fmt.Print("Введите строку: ")
	fmt.Scan(&rings)
	return rings
}

// Функция для вывода результата
func output(result int) {
	fmt.Printf("Количество стержней с 3 цветами: %d\n", result)
}

func main() {
	rings := input()

	// Карта для хранения цветов на каждом стержне
	rodColors := make(map[int]map[rune]struct{})
	
	// Инициализация карты для 10 стержней (0-9)
	for i := 0; i < 10; i++ {
		rodColors[i] = make(map[rune]struct{})
	}

	// Обработка строки колец (цвет и номер стержня)
	for i := 0; i < len(rings); i += 2 {
		color := rune(rings[i])      // Цвет кольца
		rod := int(rings[i+1] - '0') // Номер стержня
		rodColors[rod][color] = struct{}{} // Добавляем цвет к стержню
	}

	result := 0 // Счётчик стержней с 3 цветами

	// Подсчёт стержней с тремя разными цветами
	for i := 0; i < 10; i++ {
		if len(rodColors[i]) == 3 {
			result++
		}
	}

	output(result)
}

File: rings.ts
const readline = require('readline');

const rl = readline.createInterface({
input: process.stdin,
output: process.stdout
});

// Функция для ввода строки с кольцами
function input(callback: (rings: string) => void): void {
rl.question("Введите строку: ", (rings: string) => {
callback(rings);
rl.close();
});
}

// Функция для вывода результата
function output(result: number): void {
console.log("Количество стержней с 3 цветами: " + result);
}

function main(): void {
input((rings: string) => {
// Объект, где ключ - номер стержня, а значение - Set цветов
const rods: { [key: number]: Set<string> } = {};
// Инициализируем стержни
for (let i = 0; i < 10; i++) {
rods[i] = new Set<string>();
}

// Проходим по строке с кольцами и цветами
for (let i = 0; i < rings.length; i += 2) {
const color: string = rings[i]; // Цвет кольца
const rod: number = parseInt(rings[i + 1]); // Стержень кольца
rods[rod].add(color); // Добавляем цвет в соответствующий стержень
}

let result: number = 0; // Количество стержней с 3 цветами

// Считаем стержни с 3 цветами
for (let i = 0; i < 10; i++) {
if (rods[i].size === 3) {
result++;
}
}

output(result);
});
}
// Вызов основной функции
main();